File: python_backend/pins/is_pin.py
from custom_except import *
from getters.get_pathdata_dict import get_pathdata_dict


def is_pin(path_data):
    """return true if the path contains a pinned piece, otherwise false 
    
        path_data: see class _PathData
    """
    if len(path_data.statuses) < 2:
        return False
    if path_data.statuses[0] != 'f':
        return False
    if path_data.statuses[1] != 'e':
        return False
    if not path_data.piece_matches_path[1]:
        return False

    return True


if __name__ == '__main__':
    board =\
        {(1, 1): 'BB1', (1, 2): '#', (1, 3): '#', (1, 4): 'BQ3', (1, 5): '#', (1, 6): '#', (1, 7): 'BQ4', (1, 8): '#',
         (2, 1): '#', (2, 2): '#', (2, 3): '#', (2, 4): '#', (2, 5): '#', (2, 6): '#', (2, 7): '#', (2, 8): '#',
         (3, 1): '#', (3, 2): '#', (3, 3): 'WN1', (3, 4): 'WQ1', (3, 5): '#', (3, 6): '#', (3, 7): '#', (3, 8): '#',
         (4, 1): 'BQ1', (4, 2): 'WR1', (4, 3): 'WP1', (4, 4): 'WK1', (4, 5): '#', (4, 6): 'WP2', (4, 7): 'BR2',
         (4, 8): 'BK1',
         (5, 1): '#', (5, 2): '#', (5, 3): '#', (5, 4): '#', (5, 5): '#', (5, 6): '#', (5, 7): '#', (5, 8): '#',
         (6, 1): '#', (6, 2): 'WR2', (6, 3): '#', (6, 4): 'WQ2', (6, 5): '#', (6, 6): 'WP3', (6, 7): '#', (6, 8): '#',
         (7, 1): 'BQ2', (7, 2): '#', (7, 3): '#', (7, 4): 'WB1', (7, 5): '#', (7, 6): '#', (7, 7): 'WP4', (7, 8): '#',
         (8, 1): '#', (8, 2): '#', (8, 3): '#', (8, 4): 'BR1', (8, 5): '#', (8, 6): '#', (8, 7): '#', (8, 8): 'BB2'}

    # test 1:
    print('test 1:')
    path_dict = get_pathdata_dict((4, 4), board, 'W')
    for path in path_dict.values():
        print("{}, {}".format(path.coord_path, is_pin(path)))
    print('\n')

    board =\
    {(1, 8): 'BR1', (2, 8): 'BN1', (3, 8): 'BB1', (4, 8): 'BQ1', (5, 8): 'BK1', (6, 8): 'BB2', (7, 8): 'BN2', (8, 8): 'BR2',
    (1, 7): 'BP1', (2, 7): 'BP2', (3, 7): 'BP3', (4, 7): 'BP4', (5, 7): 'BP5', (6, 7): 'BP6', (7, 7): 'BP7', (8, 7): 'BP8',
    (1, 6): '#', (2, 6): '#', (3, 6): '#', (4, 6): '#', (5, 6): '#', (6, 6): '#', (7, 6): '#', (8, 6): '#',
    (1, 5): '#', (2, 5): '#', (3, 5): '#', (4, 5): '#', (5, 5): '#', (6, 5): '#', (7, 5): '#', (8, 5): '#',
    (1, 4): '#', (2, 4): '#', (3, 4): '#', (4, 4): '#', (5, 4): '#', (6, 4): '#', (7, 4): '#', (8, 4): '#',
    (1, 3): '#', (2, 3): '#', (3, 3): '#', (4, 3): '#', (5, 3): '#', (6, 3): '#', (7, 3): '#', (8, 3): '#',
    (1, 2): 'WP1', (2, 2): 'WP2', (3, 2): 'WP3', (4, 2): 'WP4', (5, 2): 'WP5', (6, 2): 'WP6', (7, 2): 'WP7', (8, 2): 'WP8',
    (1, 1): 'WR1', (2, 1): 'WN1', (3, 1): 'WB1', (4, 1): 'WQ1', (5, 1): 'WK1', (6, 1): 'WB2', (7, 1): 'WN2', (8, 1): 'WR2'}

    # test 2:
    print('test 1:')
    path_dict = get_pathdata_dict((4, 4), board, 'W')
    for path in path_dict.values():
        print("{}, {}".format(path, is_pin(path)))
    print('\n')

File: node.js/pins/isPin.js
const mapListXyToRf = require("../coordType/mapListXyToRf")
const printBoard = require("../printers/printBoard")


function isPin(pathData) {
    /**return true if the path contains a pinned piece, otherwise false 
        see class PathData
    */
    if (pathData.statuses.length < 2) {
        return false
    }
    if (pathData.statuses[0] != 'f') {
        return false
    }
    if (pathData.statuses[1] != 'e') {
        return false
    }
    if (! pathData.pieceMatchesPath[1]) {
        return false
    }
    return true
}

module.exports = isPin;

if (require.main === module) {

    var getPathdataDict  = require("../pathsInfo/top/getpathDataDict");
    const getStandardPieceDefs = require("../testObjects/getStandardPieceDefs");
    const sampleBoardDicts = require("../testObjects/sampleBoardDicts");
    const getStandardIdDict = require("../testObjects/getStandardIdDict");   

    var idDict = getStandardIdDict()
    var pieceDefs = getStandardPieceDefs()
    var pins = []
    var nonPins = []

    // test 1:
    pins = []
    nonPins = []
    console.log('test 1:')
    board = sampleBoardDicts["check_example3"]
    pathDict = getPathdataDict(board, [4, 4], 'W', pieceDefs, idDict)
    for (var direction of Object.keys(pathDict)) {
        if (isPin(pathDict[direction]))
            pins.push(...mapListXyToRf(pathDict[direction].coordPath))
        else
            nonPins.push(...mapListXyToRf(pathDict[direction].coordPath))
    }
    printBoard(board, "pins: red, non-pins: green", pins, nonPins, ["d4"])
    console.log('\n')


    // test 2:
    pins = []
    nonPins = []
    console.log('test 2:')
    var board = sampleBoardDicts["pinned_ex1"];
    var pathDict = getPathdataDict(board, [4, 7], 'W', pieceDefs, idDict)
    for (var direction of Object.keys(pathDict)) {
        if (isPin(pathDict[direction]))
            pins.push(...mapListXyToRf(pathDict[direction].coordPath))
        else
            nonPins.push(...mapListXyToRf(pathDict[direction].coordPath))
    }
    printBoard(board, "pins: red, non-pins: green", pins, nonPins, ["d7"])
    console.log('\n')

    // test 3:
    pins = []
    nonPins = []
    console.log('test 3:')
    board = sampleBoardDicts["pinned_ex2"];
    var pathDict = getPathdataDict(board, [4, 5], 'W', pieceDefs, idDict)
    for (var direction of Object.keys(pathDict)) {
        if (isPin(pathDict[direction]))
            pins.push(...mapListXyToRf(pathDict[direction].coordPath))
        else
            nonPins.push(...mapListXyToRf(pathDict[direction].coordPath))
    }
    printBoard(board, "pins: red, non-pins: green", pins, nonPins, ["d5"])
    console.log('\n')


}